### PR TITLE
[5.7] Feature: Eloquent Trait Initializers

### PR DESCRIPTION
**Pull request by ralphschindler**
(This is an alternative to #24445.)

Essentially, this introduces the opportunity for traits to call an initialization method whenever an eloquent model is new'd up without the impact of using the `Dispatcher`.  There is 0 performance impact on default models and existing usages.  The performance impact for 1000 objects returned from the database that use a trait that has an initializer is a mere `2ms`, near 0%.

To use this feature, an Eloquent trait will implement a method named `initialize<TraitName>` (this is the same pattern used for static method of booting a model.

The use case for this feature is that there are a couple of areas in eloquent that still make it difficult to write extensions that can handle complex attributes.  (By "complex attributes", I mean non-natively-known attribute types that can be expressed as objects until serialization, like how datetime objects are handled). Take for example there were an attribute of type `Coordinate`, and one wishes to be able to achieve the following workflow:

```php
class Coordinate {} // assume exists as a 3rd party value object
trait HasGeoTrait { // assume exists as a 3rd party Eloquent trait
  public function initializeHasGeoTrait() {
    // this will be called on $this for every new object.
  }
}
class Place extends Model { uses 3rdParty/HasGeoTrait; ...} // assume this kind of model

$place = Place::find(1234);
$place->coordinate->setLatLong(30.01, -90.10);

// or

$place = new Place;
$place->coordinate->setLatLong(30.01, -90.10);
```

(Separately, could this be a 5.6 feature? If so I can rebase against that branch.)

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -190,10 +199,24 @@ protected static function bootTraits()
...
             if (method_exists($class, $method = 'boot'.class_basename($trait))) {
                 forward_static_call([$class, $method]);
             }
+            if (method_exists($class, $method = 'initialize'.class_basename($trait))) {
+                static::$traitInitializers[$class][] = $method;
+            }
+        }
+    }
+
+    protected function initializeTraits()
```
[the remark this answers is not in the document]
Comment (ralphschindler) [reply src/Illuminate/Database/Eloquent/Model.php]: will add. thanks!

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -190,10 +199,24 @@ protected static function bootTraits()
...
+    {
+        $class = static::class;
```
[the remark this answers is not in the document]
Comment (ralphschindler) [reply src/Illuminate/Database/Eloquent/Model.php]: sound suggestion!

File: src/Illuminate/Database/Eloquent/Model.php [hunk 3]
```diff
@@ -190,10 +199,27 @@ protected static function bootTraits()
...
                 forward_static_call([$class, $method]);
             }
+            if (method_exists($class, $method = 'initialize'.class_basename($trait))) {
+                static::$traitInitializers[$class][] = $method;
+            }
+        }
+    }
+
+    /**
+     * Initialize any initializable traits on the model.
```
Comment (Kyslik) [inline src/Illuminate/Database/Eloquent/Model.php]: Perhaps stick to the boot doc:

> Initialize all of the initializable traits on the model.

Comment (ralphschindler) [reply src/Illuminate/Database/Eloquent/Model.php]: I don't follow completely what you are suggesting. :)

Comment (Kyslik) [reply src/Illuminate/Database/Eloquent/Model.php]: https://github.com/ralphschindler/laravel-framework/blob/0cab06b58765a70dcb7af1e4758e538190b375ee/src/Illuminate/Database/Eloquent/Model.php#L193-L198

It's whatever; just OCD I don't know.

**Pull request merged**